### ISSUE TITLE
metadata: title not updated

### ISSUE THREAD

**Issue by TurboGit** [darktable-org/darktable#5826]
**Describe the bug**
The title (and possibly other metadata) are not properly updated from XMP.

**To Reproduce**
1. On lighttable select one image
2. Add title "mytitle"
3. Make sure the option to check for updated XMP on disk is activated
3. Close darktable'
4. With one editor change "mytitle" to "mytitleNEW" for example
5. Start darktable
6. When asked to update XMP do so, select the picture and update dt library from XMP (left button)
7. See error : the title has been removed from image

**Expected behavior**
The title should have been updated

**Platform (please complete the following information):**
 - Darktable Version: master
 - OS: GNU/Debian
 
**Additional context**
Maybe related to #5722.

**Comment by TurboGit** [comment]
@phweyland : Can you have a look ? I have made the reproducer as simple as possible.

**Comment by TurboGit** [comment]
Just tested, the same issue arise for all metadata, not only title.

**Comment by TurboGit** [comment]
We have a race condition somewhere and this may be the issue. Another test I did on 225 pictures.
- Select all pictures
- Assign tag `tagtoreplace`
- Quit dt
- On the command line mass change the tag:
`sed -i -s 's/tagtoreplace/tagtoreplaceNEW/g' *.xmp`
- Start dt, when asked select all pictures and update db from XMP

Result is that some images have the `tagtoreplaceNEW` and some others `tagtoreplace`.

**Comment by phweyland** [comment]
> @phweyland : Can you have a look ? I have made the reproducer as simple as possible.

For sure

**Comment by phweyland** [comment]
Hi Pascal, the two scenarios above work for me with the current master.
I've made the test with a fresh db, imported some images and run your scenarios.
No change in the settings except "look for updated xmp files on startup".
...

**Comment by phweyland** [comment]
I've repeated the second scenario with 367 images both for metadata and tags (metatoreplace and tagtoreplace) and everything is still fine.

**Comment by TurboGit** [comment]
Good news... But worrisome as I can consistently reproduce the first scenario on my side.

After having removed all XMP and imported my collection I cannot reproduce the second scenario. So let's forget about this one for now, but again it may be triggered by some context issue...

**Comment by TurboGit** [comment]
@phweyland : of course you need the option to always write an XMP sidecar to be activated. Is that the case on your side?

**Comment by TurboGit** [comment]
@phweyland : I'm really worried now... I cannot find the code to read the metadata (title, published...). And this simple scenario show the issue:

- Set title & published on an image for example.
- Remove the image from the database
- Re-import it, not a single metadtata is set

Can you reproduce this one? I'm building with current master.

**Comment by TurboGit** [comment]
This is blocking for the release as the XMP is not anymore a backup. We loose some very precious data!

**Comment by phweyland** [comment]
> @phweyland : of course you need the option to always write an XMP sidecar to be activated. Is that the case on your side?

Yes, it is. If not, "look for updated xmp files on startup" is not effective.

**Comment by phweyland** [comment]
> And this simple scenario show the issue:
> 
>     * Set title & published on an image for example.
>     * Remove the image from the database
>     * Re-import it, not a single metadtata is set
> 
> Can you reproduce this one? I'm building with current master.

This work for, I get the same metadata and tags as before.
I've rebased this morning (3.1.0+2504~g4255e5552).

**Comment by phweyland** [comment]
As I've used a fresh db, the flag "apply metadata on import" is not set. Is it the case for you too ?

**Comment by phweyland** [comment]
>  I cannot find the code to read the metadata (title, published...).

The update of metadata is done by dt_exif_xmp_read(xmp file) which is called after dt_exif_read(image file).

![image](https://user-images.githubusercontent.com/23012047/88568073-c4513f80-d00e-11ea-9045-d5710cb3cb3a.png)

**Comment by TurboGit** [comment]
@phweyland : after lot of debugging I found the issue, but I need your help for fixing.

In my import dialog options I had all metadata not selected to be imported. This also control the crawler which is updating the metadata from XMP and as discussed above we loose all of them in this case.

I suppose we want to force the import in the crawler ? And probably also when loading an XMP sidecar from the lighttable.

**Comment by phweyland** [comment]
> I suppose we want to force the import in the crawler ?

Should not, no.

**Comment by TurboGit** [comment]
Seems like a very dangerous setting to me. To me these options should ONLY be used when importing a file or folder.

**Comment by TurboGit** [comment]
As I said a user could loose all metadata very easily. No?

**Comment by phweyland** [comment]
> In my import dialog options I had all metadata not selected to be imported.

![image](https://user-images.githubusercontent.com/23012047/88568409-5c4f2900-d00f-11ea-9468-1e9f656b02d5.png)

If not checked you force import not to import those. Then it is perfectly normal not to import the metadata...
By default, all this checked, to be consistent with previous behavior.

**Comment by TurboGit** [comment]
Well those options are in the import section and here I'm ok. But changing this here could just delete all metadata from the crawler.
Maybe it is consistent... I have never used the crawler and found this issue today will trying to reproduce #5722. I was really surprise to see all my medatada lost (on a test database, so no harm on my side).

**Comment by phweyland** [comment]
> As I said a user could loose all metadata very easily. No?

Yes. I understand this. And that was the discussion I had with @houz .
That is the only way not to import data we don't care about.
Maybe the opposite logic would be safer. Something like "do not import" ?

**Comment by TurboGit** [comment]
> Maybe the opposite logic would be safer. Something like "do not import" ?

Not sure, both are ok as long as the default is import. And I'm not sure it is the case for a migration database because on my side, on my real darktable DB I have all set to off:

```
$ grep '/metadata/' ~/.config/darktable/darktablerc | grep flag
plugins/lighttable/metadata/creator_flag=0
plugins/lighttable/metadata/description_flag=0
plugins/lighttable/metadata/notes_flag=0
plugins/lighttable/metadata/publisher_flag=0
plugins/lighttable/metadata/rights_flag=0
plugins/lighttable/metadata/title_flag=0
plugins/lighttable/metadata/version name_flag=0
```
So what we need is to ensure those flags are set with DT_METADATA_FLAG_IMPORTED initially and maybe a text explaining the fact that this is also used by the crawler and the "load xmp"..

The first part initial set to DT_METADATA_FLAG_IMPORTED should be done now. The second cannot be done as we are in string freeze. Can you double check first part ?

**Comment by phweyland** [comment]
> To me these options should ONLY be used when importing a file or folder.

I would agree we could make crawler not sensitive to these flags. If someone uses digikam to add information, that's probably because he wants them too in dt.

> So what we need is to ensure those flags are set with DT_METADATA_FLAG_IMPORTED initially

This flag is set by default for all metadata. What I don't understand is why your flags are unset .. did you do it ?


> And I'm not sure it is the case for a migration database

These flags should not interfere with migration. xmp do not change and crawler should not be triggered.

> The first part initial set to DT_METADATA_FLAG_IMPORTED should be done now. The second cannot be done as we are in string freeze. Can you double check first part ?

Not sure to understand this. If for some metadata DT_METADATA_FLAG_IMPORTED is not set for a given user, we should force it ?
I'm OK to double check but I need to understand what :)

**Comment by TurboGit** [comment]
My point is if you remove from darktablerc:

plugins/lighttable/metadata/*

You start dt and you'll see that all frags are set to 0:

```
$ grep '/metadata/' ~/.config/darktable/darktablerc | grep flag
plugins/lighttable/metadata/creator_flag=0
plugins/lighttable/metadata/description_flag=0
plugins/lighttable/metadata/notes_flag=0
plugins/lighttable/metadata/publisher_flag=0
plugins/lighttable/metadata/rights_flag=0
plugins/lighttable/metadata/title_flag=0
plugins/lighttable/metadata/version name_flag=0
```

And 0 in flags means do not import (should be 4).

**Comment by phweyland** [comment]
Ooops ! this should not happen.
On my side, cleaning the .config/darktable folder, starting dt, exiting I get this:

```
$ grep '/metadata/' ~/.config/darktable/darktablerc | grep flag
plugins/lighttable/metadata/creator_flag=4
plugins/lighttable/metadata/description_flag=4
plugins/lighttable/metadata/notes_flag=4
plugins/lighttable/metadata/publisher_flag=4
plugins/lighttable/metadata/rights_flag=4
plugins/lighttable/metadata/title_flag=4
plugins/lighttable/metadata/version name_flag=5
philippe@philippe-pc:~/.config/darktable$
```

**Comment by phweyland** [comment]
> My point is if you remove from darktablerc:
> 
> `plugins/lighttable/metadata/*`

Doing this I get also:

```
$ grep '/metadata/' ~/.config/darktable/darktablerc | grep flag
plugins/lighttable/metadata/creator_flag=4
plugins/lighttable/metadata/description_flag=4
plugins/lighttable/metadata/notes_flag=4
plugins/lighttable/metadata/publisher_flag=4
plugins/lighttable/metadata/rights_flag=4
plugins/lighttable/metadata/title_flag=4
plugins/lighttable/metadata/version name_flag=5
philippe@philippe-pc:~/.config/darktable$

```

**Comment by phweyland** [comment]
The initialization is made in libs/metadata.c - init().

To explain your case, something is dealing with metadata before calling libs/metadata.c - init().

**Comment by TurboGit** [comment]
Fixed by #5827.
